Add gz-sim-yarp-basestate-system plugin to publish BaseState information

xela-95 opened the pull request:
This PR adds the `gz-sim-yarp-basestate` plugin that, as the [correspondent plugin](https://github.com/robotology/gazebo-yarp-plugins/tree/master/plugins/basestate) for gazebo classic, serves the purpose of publishing state information (base pose, velocity and acceleration) of the model base link.

Closes #59
Closes #66

xela-95 committed:
Create basestate plugin folder by copying from forcetorque

xela-95 committed:
Refactor singleton class

xela-95 committed:
Update CMakeLists

xela-95 committed:
Update Handler class member names

xela-95 committed:
Update BaseState and BaseStateDriver

xela-95 committed:
Add tutorial to test basestate plugin

xela-95 committed:
Add newline at EOF for CmakeLists files

xela-95 commented on plugins/CMakeLists.txt:
```diff
@@ -3,3 +3,4 @@ add_subdirectory(camera)
 add_subdirectory(laser)
 add_subdirectory(forcetorque)
 add_subdirectory(imu)
+add_subdirectory(basestate)
```
Thank you for the suggestion. I've enabled the setting `Files: Insert Final Newline` inside vscode editor settings. Formatted the CMakeLists files in 06a7683eb4aecb9839ee0f2417423d0c3c77ac29

xela-95 committed:
Fix basestate plugin registration in robotinterface handler

xela-95 commented:
@traversaro for this kind of system plugin, what kind of interfaces do you think I need to implement among:
- ISystemConfigure (already completed)
- ISystemPreUpdate
- ISystemUpdate
- ISystemPostUpdate
- ISystemReset

The sensor plugins already ported implement ISystemPreUpdate and ISystemPostUpdate. 

While the ISystemPostUpdate interface is needed to get data after physics simulation and update the singleton data strucure, I not sure about the ISystemPreUpdate. The documentation says it's useful to modify the state before physics runs (e.g. set control signals and perform network synchronization). In the other plugins is used as in

https://github.com/robotology/gz-sim-yarp-plugins/blob/25216f1c90ed524a30b664e7cc327b368eb2d649/plugins/forcetorque/ForceTorque.cc#L139-L148

but I'm not sure what this code does in detail.

xela-95 commented:
I've also noted that to read the base state the yarp port exporting the data is `/basestate` while for sensor plugins the port is, for example `/forcetorque/measuments`. The `/measurements` part is automatically added by some yarp interface from which it inherits? Since basestate is a "system" plugin and not a sensor do you think it is correct to have that port name exposing data?

xela-95 commented:
> While the ISystemPostUpdate interface is surely needed to get data after physics simulation and update the singleton data strucure, I'm not sure about the ISystemPreUpdate. The documentation says it's useful to modify the state before physics runs (e.g. set control signals and perform network synchronization). In the other plugins is used as in
> 
> https://github.com/robotology/gz-sim-yarp-plugins/blob/25216f1c90ed524a30b664e7cc327b368eb2d649/plugins/forcetorque/ForceTorque.cc#L139-L148
> 
> but I'm not sure what this code does in detail.
> 
> Also ISystemReset is not implemented in the plugins but I think it should in order to manage simulation resets.

@traversaro related to this I think I'm missing something on things works on Gazebo side (and I haven't found useful documentation): while in gazebo classic all the update of the data coming from gazebo was made on the callback OnUpdate, with the new Gazebo what I understand is that in the PreUpdate method we subscribe to some topic on which the Gazebo sensor is publishing data, passing a callback which has to goal of updating with the latest data a class member holding the message received, `ftMsg` in the ft plugin. 

Then, in the PostUpdate method we retrieve the data stored in this data member and store it on the singleton data structure such that the Driver class can access it and send on the yarp port.

I'm thinking on how to transpose this for Model plugins as the basestate one.
In gazebo-yarp-plugins the OnUpdate is the following:
https://github.com/robotology/gazebo-yarp-plugins/blob/cb3521f7543a5a07e2a65a82dedd2eb2ec2cea12/plugins/basestate/src/BaseStateDriver.cpp#L80-L118

xela-95 committed:
Update BaseStateDriver getLastInputStamp method

xela-95 committed:
Update BaseState to get base data with new gazebo paradigm

xela-95 committed:
Guard against cases in which data is not available

xela-95 committed:
Rename member baseLink to baseLinkEntity

xela-95 commented:
I think I've implemented the main mechanism to retrieve and store data, but I have to debug an error occurring while testing the plugin:

```
[INFO] gz-sim-yarp-basestate-system: configuration of device  basestate_plugin_device  loaded from yarpConfigurationString :  (yarpDeviceName basestate_plugin_device) (baseLink link_1) 
===================>> m_modelScopedName: model/basestate_model/link/link_1
[DEBUG] |yarp.dev.PolyDriver|gazebo_basestate| Parameters are (baseLink link_1) (device gazebo_basestate) (modelScopedName "model/basestate_model/link/link_1") (robot "model/basestate_model/link/link_1") (yarpDeviceName basestate_plugin_device)
[INFO] |yarp.dev.PolyDriver|gazebo_basestate| Created device <gazebo_basestate>. See C++ class BaseStateDriver for documentation.
[INFO] Registered YARP device with instance name: model/basestate_model/link/link_1/basestate_plugin_device
[DEBUG] Reading file basestate_nws.xml
[DEBUG] yarprobotinterface: using xml parser for DTD v3.x
[DEBUG] Reading file basestate_nws.xml
[INFO] Yarprobotinterface was started using the following enable_tags: 
[INFO] Yarprobotinterface was started using the following disable_tags: 
[DEBUG] List of all enable attributes found in the include tags: 
[DEBUG] List of all disable attributes found in the include tags: 
[DEBUG] Preprocessor complete in:  9.53674e-07 s
[DEBUG] ===========> deviceDatabaseKey:  model/basestate_model/link/link_1/basestate_plugin_device
[DEBUG] ==========> yarpDeviceName:  basestate_plugin_device
===================> Key of device 0 is basestate_plugin_device
[INFO] startup phase starting...
[INFO] Opening device basestate_nws_yarp with parameters [("robotName" = "basestate"), ("name" = "/basestate"), ("period" = "50")]
[DEBUG] |yarp.dev.PolyDriver|basestate_nws_yarp| Parameters are (device analogServer) (id basestate_nws_yarp) (name "/basestate") (period 50) (robotName basestate)
[WARNING] |yarp.devices.AnalogWrapper| The 'AnalogWrapper' device is deprecated.
[WARNING] |yarp.devices.AnalogWrapper| Possible alternatives, depending on the specific type sensor data, are:
[WARNING] |yarp.devices.AnalogWrapper| 'MultipleAnalogSensorsRemapper`+`MultipleAnalogSensorsServer`, `PoseStampedRosPublisher`, `WrenchStampedRosPublisher`,`IMURosPublisher`,etc.
[WARNING] |yarp.devices.AnalogWrapper| The old device is no longer supported, and it will be deprecated in YARP 3.7 and removed in YARP 4.
[WARNING] |yarp.devices.AnalogWrapper| Please update your scripts.
[INFO] |yarp.os.Port|/basestate/rpc:i| Port /basestate/rpc:i active at tcp://10.240.79.41:10002/
[INFO] |yarp.dev.PolyDriver|basestate_nws_yarp| Created wrapper <analogServer>. See C++ class AnalogWrapper for documentation.
[INFO] Entering action level 5 of phase startup
[INFO] Executing attach action, level 5 on device basestate_nws_yarp with parameters [("device" = "basestate_plugin_device")]
[INFO] basestate_nws_yarp is not an IWrapper. Trying IMultipleWrapper
[INFO] |yarp.os.Port|/basestate| Port /basestate active at tcp://10.240.79.41:10003/
[WARNING] BaseState data not available
[ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
[INFO] All actions for action level 5 of startup phase started. Waiting for unfinished actions.
[INFO] All actions for action level 5 of startup phase finished.
[INFO] startup phase finished.
[WARNING] BaseState data not available
[ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
[WARNING] BaseState data not available
[ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
[WARNING] BaseState data not available
[ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
[WARNING] BaseState data not available
[ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
[WARNING] BaseState data not available
[ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
[WARNING] BaseState data not available
[ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
[WARNING] BaseState data not available
[ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.

```

I think this could be due to the way in which I'm retrieving Link data:

```cpp
Link baseLink = Link(m_baseLinkEntity);

...
 // Get the pose of the origin of the link frame in the world reference frame
if (dataAvailable && baseLink.WorldPose(_ecm).has_value())
    math::Pose3d worldBasePose = baseLink.WorldPose(_ecm).value();
else
    dataAvailable = false;
```

traversaro commented on plugins/basestate/BaseStateDriver.cpp:
```diff
@@ -94,6 +94,8 @@ class yarp::dev::gzyarp::BaseStateDriver : public yarp::dev::DeviceDriver,
     yarp::os::Stamp getLastInputStamp()
     {
+        std::lock_guard<std::mutex> lock(m_baseLinkData->m_mutex);
+        m_lastTimestamp.update(m_baseLinkData->m_simTime);
```
The `yarp::os::Stamp` also have a sequence number, that is increased for any call to update (see https://yarp.it/latest/classyarp_1_1os_1_1Stamp.html#a2b836a2044d0929ad4e5a339ae0c1635). It may be strange that the sequence number tracks the number of times `getLastInputStamp()` has been called, probably it could make sense to move this `m_lastTimestamp.update` call to when `m_baseStateData.m_simTime` is updated?

xela-95 replied:
Updated in b0c3637958f9a6d029215c5aea4b1ae436763270

traversaro commented on plugins/basestate/BaseState.cc:
```diff
@@ -99,6 +97,11 @@ class BaseState : public System,
         auto model = Model(_entity);
         this->m_baseLink = model.LinkByName(_ecm, baseLinkName);
```
What if `baseLinkName` does not match any link in the model?

xela-95 replied:
Testing with a non existen link name I get the following error:

```
[Err] [EntityComponentManager.cc:1083] Trying to create a component of type [6822503852486540549] attached to entity [0], but this entity does not exist. This create component request will be ignored.
```

I will use a try block to verify that the link exists, otherwise raise an error with a log message.

xela-95 replied:
Solved in eb0ce408eb72439d69286e0305bcfec275b5d96d

traversaro commented on plugins/basestate/BaseState.cc:
```diff
@@ -99,6 +97,11 @@ class BaseState : public System,
         this->m_baseLink = model.LinkByName(_ecm, baseLinkName);
 
+        // Enable velocity computation in Gazebo
+        Link(this->m_baseLink).EnableAccelerationChecks(_ecm, true);
```
```suggestion
        Link(this->m_baseLink).EnableVelocityChecks(_ecm, true);
```

typo?

xela-95 replied:
Yep nice catch! Thank you!

xela-95 replied:
Solved in 5e0ec19563fa13c163c27bd1abc2926b8b88523c

traversaro reviewed (changes requested):
Some comments.

traversaro commented:
> I think I've implemented the main mechanism to retrieve and store data, but I have to debug an error occurring while testing the plugin:
> 
> ```
> [INFO] gz-sim-yarp-basestate-system: configuration of device  basestate_plugin_device  loaded from yarpConfigurationString :  (yarpDeviceName basestate_plugin_device) (baseLink link_1) 
> ===================>> m_modelScopedName: model/basestate_model/link/link_1
> [DEBUG] |yarp.dev.PolyDriver|gazebo_basestate| Parameters are (baseLink link_1) (device gazebo_basestate) (modelScopedName "model/basestate_model/link/link_1") (robot "model/basestate_model/link/link_1") (yarpDeviceName basestate_plugin_device)
> [INFO] |yarp.dev.PolyDriver|gazebo_basestate| Created device <gazebo_basestate>. See C++ class BaseStateDriver for documentation.
> [INFO] Registered YARP device with instance name: model/basestate_model/link/link_1/basestate_plugin_device
> [DEBUG] Reading file basestate_nws.xml
> [DEBUG] yarprobotinterface: using xml parser for DTD v3.x
> [DEBUG] Reading file basestate_nws.xml
> [INFO] Yarprobotinterface was started using the following enable_tags: 
> [INFO] Yarprobotinterface was started using the following disable_tags: 
> [DEBUG] List of all enable attributes found in the include tags: 
> [DEBUG] List of all disable attributes found in the include tags: 
> [DEBUG] Preprocessor complete in:  9.53674e-07 s
> [DEBUG] ===========> deviceDatabaseKey:  model/basestate_model/link/link_1/basestate_plugin_device
> [DEBUG] ==========> yarpDeviceName:  basestate_plugin_device
> ===================> Key of device 0 is basestate_plugin_device
> [INFO] startup phase starting...
> [INFO] Opening device basestate_nws_yarp with parameters [("robotName" = "basestate"), ("name" = "/basestate"), ("period" = "50")]
> [DEBUG] |yarp.dev.PolyDriver|basestate_nws_yarp| Parameters are (device analogServer) (id basestate_nws_yarp) (name "/basestate") (period 50) (robotName basestate)
> [WARNING] |yarp.devices.AnalogWrapper| The 'AnalogWrapper' device is deprecated.
> [WARNING] |yarp.devices.AnalogWrapper| Possible alternatives, depending on the specific type sensor data, are:
> [WARNING] |yarp.devices.AnalogWrapper| 'MultipleAnalogSensorsRemapper`+`MultipleAnalogSensorsServer`, `PoseStampedRosPublisher`, `WrenchStampedRosPublisher`,`IMURosPublisher`,etc.
> [WARNING] |yarp.devices.AnalogWrapper| The old device is no longer supported, and it will be deprecated in YARP 3.7 and removed in YARP 4.
> [WARNING] |yarp.devices.AnalogWrapper| Please update your scripts.
> [INFO] |yarp.os.Port|/basestate/rpc:i| Port /basestate/rpc:i active at tcp://10.240.79.41:10002/
> [INFO] |yarp.dev.PolyDriver|basestate_nws_yarp| Created wrapper <analogServer>. See C++ class AnalogWrapper for documentation.
> [INFO] Entering action level 5 of phase startup
> [INFO] Executing attach action, level 5 on device basestate_nws_yarp with parameters [("device" = "basestate_plugin_device")]
> [INFO] basestate_nws_yarp is not an IWrapper. Trying IMultipleWrapper
> [INFO] |yarp.os.Port|/basestate| Port /basestate active at tcp://10.240.79.41:10003/
> [WARNING] BaseState data not available
> [ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
> [INFO] All actions for action level 5 of startup phase started. Waiting for unfinished actions.
> [INFO] All actions for action level 5 of startup phase finished.
> [INFO] startup phase finished.
> [WARNING] BaseState data not available
> [ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
> [WARNING] BaseState data not available
> [ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
> [WARNING] BaseState data not available
> [ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
> [WARNING] BaseState data not available
> [ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
> [WARNING] BaseState data not available
> [ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
> [WARNING] BaseState data not available
> [ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
> [WARNING] BaseState data not available
> [ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
> ```
> 
> I think this could be due to the way in which I'm retrieving Link data:
> 
> ```c++
> Link baseLink = Link(m_baseLinkEntity);
> 
> ...
>  // Get the pose of the origin of the link frame in the world reference frame
> if (dataAvailable && baseLink.WorldPose(_ecm).has_value())
>     math::Pose3d worldBasePose = baseLink.WorldPose(_ecm).value();
> else
>     dataAvailable = false;
> ```

I think one of my comments is related to that.

xela-95 committed:
Fix bug in BaseState.cc

xela-95 committed:
Check whether the base link name exists

xela-95 committed:
Change the update of the last timestamp

xela-95 committed:
Fix how Link class is used to retrieve data

xela-95 committed:
Add EOL to RobotInterface.cc

xela-95 committed:
Update tutorial

xela-95 committed:
Refactorings

- Update the code style of the plugin to be more self-explainable and adherent to the repo conventions;
- This commit refactors the base state handler in the `singleton-basestate` plugin. The `Handler.hh` file has been removed and replaced with `BaseStateDataSingleton.hh`. The `BaseStateDataSingleton` class now handles the management of base state data, including setting, getting, and removing base state data.

xela-95 commented:
@traversaro I updated the code addressing your comments and trying to refactor the code to be more explainable and avoiding useless variables. I tested the code using the tutorial and it worked. However there is still the log of the analog sensor when gazebo is launched but the simulation is not yet started (since the base state data is not yet available):

```
[WARNING] BaseState data not available
[ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
```

How do you suggest to address it?

traversaro commented:
> @traversaro I updated the code addressing your comments and trying to refactor the code to be more explainable and avoiding useless variables. I tested the code using the tutorial and it worked. However there is still the log of the analog sensor when gazebo is launched but the simulation is not yet started (since the base state data is not yet available):
> 
> ```
> [WARNING] BaseState data not available
> [ERROR] |yarp.devices.AnalogWrapper| AnalogWrapper: : Sensor returned error with code 1.
> ```
> 
> How do you suggest to address it?

This is actually an error in `AnalogWrapper`, that however is a deprecated piece of software that it does not make sense to fix. We can just ignore, alternativly we could also block (with a timeout) on the first `read` method, to not return until a first good measure is available.

Something like adding a `m_firstDataAvailable` boolean attribute to `yarp::dev::gzyarp::BaseStateDriver`, that at the beginning is false, and modify read as:

~~~cxx
    int read(yarp::sig::Vector& out)
    {
        if (out.size() != YarpBaseStateChannelsNumber)
        {
            out.resize(YarpBaseStateChannelsNumber);
        }


        bool dataAvailable = false;
        {
            std::lock_guard<std::mutex> lock(m_baseLinkData->mutex);
            dataAvailable = m_baseLinkData->dataAvailable;
        }

        // This code just get executed in the first read
        if (!m_firstDataAvailable)
        {
             if (dataAvailable)
             {
                 m_firstDataAvailable = true;
             } else {
                 double waitingTimeInS = 0.0;
                 double timeoutInFirstWaitInS = 1.0;
                 double checkPeriodInS = 0.01;
                 
                 while (waitingTimeInS <= timeoutInFirstWaitInS && !dataAvailable )
                 {
                     std::this_thread::sleep_for(std::chrono::milliseconds(checkPeriodInS*1000));
                     {
                          std::lock_guard<std::mutex> lock(m_baseLinkData->mutex);
                          dataAvailable = m_baseLinkData->dataAvailable;
                     }
                 }

                 if (!dataAvailable)
                 {
                     // priint warning and return error value
                 }
             }
        }


        {
            std::lock_guard<std::mutex> lock(m_baseLinkData->mutex);

            for (size_t i = 0; i < YarpBaseStateChannelsNumber; i++)
            {
               out[i] = m_baseLinkData->data[i];
            }
        }

        return AS_OK;
    }
~~~

(I also removed an unnecessary buffer).

traversaro commented on plugins/basestate/BaseState.cc:
```diff
@@ -0,0 +1,243 @@
+                           EntityComponentManager& _ecm,
+                           EventManager& /*_eventMgr*/) override
+    {
+        yarp::os::Network::init();
```
I know that this is everywhere in the plugins, but like it does not make sense to manually lock and unlock mutexes, similarly it does not make sense to manually call `yarp::os::Network::init()` and `        yarp::os::Network::fini();`, you can just add an attribute `yarp::os::Network m_yarpNetwork;` to the class, the `init` and `fini` will be called in its constructor and destructor (this is typically called RAII).

xela-95 replied:
Solved in bc9dde81de292f5c1656733de1339015b8c3dddb

traversaro commented on plugins/basestate/BaseState.cc:
```diff
@@ -0,0 +1,243 @@
+                           EventManager& /*_eventMgr*/) override
+    {
+        yarp::os::Network::init();
+        if (!yarp::os::Network::checkNetwork())
+        {
+            yError() << "Yarp network does not seem to be available, is the yarpserver running?";
+            return;
+        }
```
```suggestion
```

Again, also this is everyhere, so it is not your fault. However, this comes from another era, in which all network wrapper servers were using the YARP network. As now wrapper could just use other stuff (like ROS2), we can avoid to check if network is available. Network wrapper servers that use YARP ports will care if a YARP network is actually available.

traversaro commented on plugins/basestate/BaseState.cc:
```diff
@@ -0,0 +1,243 @@
+                                                                                   "basestate",
+                                                                                   netWrapper
+                                                                                       .c_str(),
+                                                                                   "BaseStateDrive"
+                                                                                   "r"));
```
This is quite hard to read, and it is not obvious that the two strings are actually concatenated by the compiler. Do you have any idea of how to make this more readable?

traversaro replied:
Note that you can locally disable clang-format with https://clang.llvm.org/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code .

xela-95 replied:
Solved in abb966b648750631dd1496e5e4cfb4bc831d1aa5 , I defined an alias for the long type in order to have this line more readable.

traversaro commented on plugins/basestate/BaseState.cc:
```diff
@@ -0,0 +1,243 @@
+        ::yarp::os::Property driver_properties;
+
+        bool wipe = false;
+        if (_sdf->HasElement("yarpConfigurationString"))
```
I just noticed that we are supporting `yarpConfigurationFile` only in the laser plugin. Probably it could make sense to harmonize this across plugins (anyhow it is not blocking this PR): https://github.com/robotology/gz-sim-yarp-plugins/issues/68 .

traversaro commented on plugins/basestate/BaseState.cc:
```diff
@@ -0,0 +1,243 @@
+        m_deviceScopedName
+            = m_baseLinkScopedName + "/" + driver_properties.find("yarpDeviceName").asString();
+
+        if (!Handler::getHandler()->setDevice(m_deviceScopedName, &m_baseStateDriver))
```
I opened this https://github.com/robotology/gz-sim-yarp-plugins/issues/69 to avoid forgetting.

xela-95 replied:
Ok thanks!

traversaro commented on plugins/basestate/BaseState.cc:
```diff
@@ -0,0 +1,243 @@
+        math::Vector3d worldBaseAngAcc;
+
+        // Get the pose of the origin of the link frame in the world reference frame
+        if (dataAvailable && m_baseLink.WorldPose(_ecm).has_value())
+            worldBasePose = m_baseLink.WorldPose(_ecm).value();
+        else
+            dataAvailable = false;
```
This is just something that is matter of taste, so do as you prefer. But I typically find if/else bodies without any bracket to be quite confusing, see https://stackoverflow.com/a/2125078 for the reason why.

xela-95 replied:
Sure, I used to do it, but I agree that is less future-proof (btw interesting article!) I will update it.

xela-95 replied:
Solved in cc3bcc7ad30d2c1bd32c97c021dc388188fdde17

traversaro commented on plugins/basestate/BaseState.cc:
```diff
@@ -0,0 +1,243 @@
+        else
+            dataAvailable = false;
+
+        std::lock_guard<std::mutex> lock(m_baseStateData.mutex);
```
As for if/else, it may be more clear if we give this lockguard its own scope, otherwise people in the future may add a line at the end of this method, without realizing that they are actually doing stuff in the scope of this lock_guard.

xela-95 replied:
Sounds good, solved in 821498f80ef0a1645033aa19cdc807b19add48a8 also in other places inside the plugin.

traversaro commented on plugins/basestate/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+project(basestate)
```
```suggestion
project(basestate)
```

I think we can remove this.

xela-95 replied:
We can remove this since it is not referenced anywhere right?

traversaro commented on plugins/basestate/singleton-basestate/BaseStateDataSingleton.cc:
```diff
@@ -0,0 +1,95 @@
+BaseStateData*
+BaseStateDataSingleton::getBaseStateData(const std::string& _baseLinkScopedName) const
+{
+    BaseStateData* tmp;
```
```suggestion
    BaseStateData* tmp;
```

tmp is quite a generic name, perhaps we can use retptr or something more meaningful?

xela-95 replied:
Changed here a97121d00ee27a65dac4c0548b3ec6ba9acbc407

traversaro commented on plugins/basestate/singleton-basestate/BaseStateDataSingleton.cc:
```diff
@@ -0,0 +1,95 @@
+    } else
+    {
+        yError() << "Base link was not found: " << _baseLinkScopedName;
+        tmp = NULL;
```
```suggestion
        tmp = nullptr;
```

In C++ version >=11, always use `nullptr`.

traversaro commented on tutorial/basestate/README.md:
```diff
@@ -0,0 +1,91 @@
+
+As you can observe from the basestate plugin output, the base link frame starts at 0.2m height on Z direction (the 3rd value printed on each line) and when the simulation runs the cube falls on the ground; the Z value then remains constant to ~0.05m.
+
+> [!NOTE]
+> If you are using Linux on WSLg and you have the following error:
+>
+> ~~~
+> OGRE EXCEPTION(9:UnimplementedException):  in GL3PlusTextureGpu::copyTo at ./RenderSystems/GL3Plus/src/OgreGL3PlusTextureGpu.cpp (line 685)
+> ~~~
+>
+> try forcing software rendering:
+>
+> ~~~
+> export LIBGL_ALWAYS_SOFTWARE=1
+> ~~~
```
```suggestion
```

Not necessary anymore, see https://github.com/robotology/gz-sim-yarp-plugins/issues/40#issuecomment-1862555770 .

xela-95 replied:
Thanks! Should we remove this note from all the tutorials then?

traversaro replied:
Sure, but better not do to it here, just use another PR. We can track this in https://github.com/robotology/gz-sim-yarp-plugins/issues/40#issuecomment-1862555770 .

traversaro reviewed (approved):
Minor changes, most style, do not block the merge.

xela-95 committed:
Update tutorial/basestate/README.md

Co-authored-by: Silvio Traversaro <silvio@traversaro.it>

xela-95 committed:
Update plugins/basestate/singleton-basestate/BaseStateDataSingleton.cc

Use nullptr instead of NULL

Co-authored-by: Silvio Traversaro <silvio@traversaro.it>

xela-95 committed:
Update plugins/basestate/CMakeLists.txt

Co-authored-by: Silvio Traversaro <silvio@traversaro.it>

xela-95 committed:
Rename local variable in getBaseStateData

xela-95 committed:
Uniform if else convention to use always use {} brackets

xela-95 committed:
Fix missing bracket in else clause

(@traversaro was right!)

xela-95 committed:
Delimit lock_guard scope in basestate plugin

xela-95 committed:
Define alias to avoid ugly formatting in BaseState.cc

xela-95 committed:
Add yarp network member to BaseState and delegate to it network initialization and finalization phases

xela-95 committed:
Update plugins/basestate/BaseState.cc

Avoid to check yarp network availability in this class.

Co-authored-by: Silvio Traversaro <silvio@traversaro.it>

xela-95 commented:
@traversaro all comments related to this plugin have been addressed. Merging.
